module/signer.go

```diff
@@ -29,6 +32,33 @@ type ThresholdSigner interface {
 	Reconstruct(size uint, shares []crypto.Signature, indices []uint) (crypto.Signature, error)
 }
 
+var (
+	// DKGIncompleteError indicates that the node did not manage to obtain beacon keys from DKG at a certain view
+	DKGIncompleteError = errors.New("incomplete dkg")
+)
+
+// TODO: to be replaced by RandomBeaconSignerStore
+// ThresholdSignerStore returns the threshold signer object by view
 type ThresholdSignerStore interface {
+	// It returns:
+	//  - (signer, nil) if the node has beacon keys in the epoch of the view
+	//  - (nil, DKGIncompleteError) if the node doesn't have beacon keys in the epoch of the view
+	//  - (nil, error) if there is any exception
 	GetThresholdSigner(view uint64) (ThresholdSigner, error)
 }
+
+// MsgSigner signs a given message and produces a signature
+// TODO: could be renamed to Signer once the original Signer is replaced with MsgSigner
+type MsgSigner interface {
+	Sign(msg []byte) (crypto.Signature, error)
+}
+
+// RandomBeaconSignerStore returns the signer for the given view, and it internally caches
+// signer objects by epoch. And epoch is determined by view.
+type RandomBeaconSignerStore interface {
+	// It returns:
+	//  - (signer, nil) if the node has beacon keys in the epoch of the view
+	//  - (nil, DKGIncompleteError) if the node doesn't have beacon keys in the epoch of the view
+	//  - (nil, error) if there is any exception
+	GetSigner(view uint64) (MsgSigner, error)
+}
```

Comment:
I will just be one MsgSigner per epoch, rather than one MsgSigner per view. 

If we return the private key, who should hold the hasher? 

I'm planning to make a Factory method, such that it ensures the same hasher will be used by both SignerStore and Verifier. 

Something like:
```
// for shared code
randomBeaconFactory = makeFactoryWithTag(RandomBeaconTag) // create hasher

// for signer
randomBeaconSignerStore = randomBeaconFactory.CreateSigerStore()
...

// for verifier
randomBeaconVerifier = randomBeaconFactory.CreateVerifier()
```

Reply:
You are right that one extra hasher per epoch is a negligible overhead. 
My suggestion is more related to the clarity of the design : a store for different data per epoch can be limited to what is really different every epoch. In this case it is just the private key.